sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -322,7 +322,12 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
    */
   private def getByteArrayRdd(
       n: Int = -1, takeFromEnd: Boolean = false): RDD[(Long, Array[Byte])] = {
-    execute().mapPartitionsInternal { iter =>
+    val rdd = if (supportsColumnar) {
+      ColumnarToRowExec(this).execute()
```

Comment:
Sorry for the late review, but I have a question about this change.

Actually, `SparkPlan.execute` has the same issue, and we rely on the framework to add `ColumnarToRowExec`, so that we never call `execute` on a spark plan that is columnar. Why do we take a different approach for `SparkPlan.executeCollect`? Is it possible to guarantee that we never call `executeCollect` on a columnar spark plan, with the existing columnar framework?

Reply:
From my experience that is what happens already. executeCollect will be called on the `ColumnarToRowExec`, not on anything that supports columnar processing.  But I don't know what issues others have run into. For us one of the main issues comes up with operators like `CollectLimitExec` which want their `executeCollect` to be called so they can limit the number of tasks that have to run to produce a result. If the `ColumnarToRowExec` is inserted after a columnar version of `CollectLimitExec`, like it is done for this patch, then execute is called and you get less desirable behavior.

Reply:
`execute` and `executeColumnar` are APIs internally used for query execution, but `executeCollect` is somehow different as it works more like a human-facing collect API, though it is not actually end-user facing one. But it collects internal rows and decodes them, it's convenient one to use to examine the output of arbitrary SparkPlan.

In other words, I can't use `SparkPlan.execute` or `SparkPlan.executeColumnar` to collect meaningful rows (readable) but I think `executeCollect` should be able for that. Previously it is only for non-columnar plan, though.

The framework is to guarantee what you said for the entire query. But there is no `ColumnarToRowExec` after each columnar plan (as some might be continuously columnar). During development, if developers want to examine arbitrary columnar plan, the folks may need to convert it manually. I feel it is verbose for every time to do it again.

Reply:
I'm a bit uncomfortable with creating query plans on the fly in an execution code path. People may have to apply some final transformation before running a physical plan, but now there is no chance to do it. How about this
```
class SparkPlan {
  def toRowBased: SparkPlan = if (supportsColumnar) ColumnarToRowExec(this) else this
}
```

Then during debugging we can do `plan.toRowBased.executeCollect()`

Reply:
Looks good for my purpose. I will create a followup for it. Thanks.